File: Leetcode/125. Valid Palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        # Preprocess the string: remove non-alphanumeric and convert to lowercase
        filtered = ''.join(char.lower() for char in s if char.isalnum())

        # Define a recursive helper function
        def check(left: int, right: int) -> bool:
            if left >= right:
                return True
            if filtered[left] != filtered[right]:
                return False
            return check(left + 1, right - 1)

        return check(0, len(filtered) - 1)

File: Leetcode/125. Valid Palindrome.java
class Solution {
    public boolean isPalindrome(String s) {
        int left = 0;
        int right = s.length() - 1;

        while (left < right) {
            while (left < right && !isAlphaNum(s.charAt(left))) {
                left++;
            }
            while (left < right && !isAlphaNum(s.charAt(right))) {
                right--;
            }
            if (Character.toLowerCase(s.charAt(left)) != Character.toLowerCase(s.charAt(right))) {
                return false;
            }
            left++;
            right--;
        }
        return true;
    }

    private boolean isAlphaNum(char n) {
        return (n >= 'a' && n <= 'z') || (n >= 'A' && n <= 'Z') || (n >= '0' && n <= '9');
    }
}
